keps/sig-scheduling/3633-matchlabelkeys-to-podaffinity/README.md

```diff
@@ -0,0 +1,930 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes introducing a complementary field `MatchLabelKeys` to `PodAffinityTerm`.
+This enables users to finely control the scope where Pods are expected to co-exist (PodAffinity)
+or not (PodAntiAffinity), on top of the existing `LabelSelector`.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+During a workload's rolling upgrade, depending on its `upgradeStrategy`, old and new versions of
+Pods may co-exist in the cluster. 
+As scheduler cannot distinguish "old" from "new", it cannot properly honor the API semantics of 
+PodAffinity and PodAntiAffiniy during the upgrade. 
+In the worse case, new version of Pod cannot be scheduled if it's a saturated cluster: [1], [2], [3].
+
+[1]: https://github.com/kubernetes/kubernetes/issues/56539#issuecomment-449757010
+[2]: https://github.com/kubernetes/kubernetes/issues/90151
+[3]: https://github.com/kubernetes/kubernetes/issues/103584
+
+On the other hand, on an idle cluster, this can cause the scheduling result sub-optimal because
+some qualifying Nodes are filtered out incorrectly.
+
+From the same motivation, we've introduced `MatchLabelKeys` in Pod Topology Spread.
```

Comment:
The same issue applies to other scheduling directives as well. For example, `MatchLabelKeys` was introduced in topologyConstaint in  [KEP-3243: Respect PodTopologySpread after rolling upgrades](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades).